Libraries/LibJS/Runtime/ArrayPrototype.cpp

```diff
@@ -347,6 +348,54 @@ Value ArrayPrototype::index_of(Interpreter& interpreter)
     return Value(-1);
 }
 
+Value ArrayPrototype::reduce(Interpreter& interpreter)
+{
+    auto this_value = interpreter.this_value();
+    auto* this_object = this_value.to_object(interpreter);
+    if (!this_object)
+        return {};
+
+    auto initial_length = get_length(interpreter, *this_object);
+    if (interpreter.exception())
+        return {};
+
+    size_t start = 1;
+
+    auto accumulator = js_undefined();
+    if (interpreter.argument_count() > 1) {
+        accumulator = interpreter.argument(1);
+        start = 0;
+    } else if (initial_length >= 1) {
+        accumulator = this_object->get_by_index(0);
+    } else {
+        interpreter.throw_exception<TypeError>(String::format("Reduce of empty array with no initial value"));
+        return {};
+    }
+
+    auto* callback_function = callback_from_args(interpreter, "reduce");
+    if (!callback_function)
+        return {};
+
+    for (size_t i = start; i < initial_length; ++i) {
+        auto value = this_object->get_by_index(i);
```

Comment:
Missing exception check:

```suggestion
        auto value = this_object->get_by_index(i);
        if (interpreter.exception())
            return {};
```

Reply:
Seems like those two places would be the only calls to `get_by_index` followed by an exception check. I'm not sure if this method can even throw a JS exception.

Reply:
It definitely can throw - `get_by_index()` has a code path for negative values that just calls regular `get()` which you can very easily blow up using a property getter function for example.

https://github.com/SerenityOS/serenity/blob/45dfa094e9ab88c4af833674e85a93b704c79988/Libraries/LibJS/Runtime/Object.cpp#L371-L372

Now since here the index is guaranteed to be >= 0 we could leave it out RIGHT NOW but that's actually due to a missing piece in the implementation - getters for positive indices will be possible too, see https://github.com/SerenityOS/serenity/issues/2330.

> Seems like those two places would be the only calls to get_by_index followed by an exception check.

There are others:

https://github.com/SerenityOS/serenity/blob/040c75a3cc8ae4ac87bcc03bf741b47bc688d2a0/Libraries/LibJS/Runtime/ArrayPrototype.cpp#L224-L226

https://github.com/SerenityOS/serenity/blob/040c75a3cc8ae4ac87bcc03bf741b47bc688d2a0/Libraries/LibJS/Runtime/ArrayPrototype.cpp#L275-L277

`for_each_item` is missing one.